### PR TITLE
[WIP] Fix problem where tree search is truncated incorrectly.

### LINKED ISSUE #200: SBT search functions currently calculate containment
The SBT tree search is based on maximizing `count_common`, not on maximizing the MinHash similarity.  This is now different from the behavior of `sourmash search` and `MinHash.compare()`.

### LINKED ISSUE #257: add a '--traverse-dir' argument to search and gather
goal would be to find relevant signatures within directory.

### LINKED ISSUE #281: limit 'search --best-only' output to a single result
see https://github.com/dib-lab/sourmash/issues/275#issuecomment-306824947

### PR DESCRIPTION
**Pull request by ctb**
Fixes #200, where tree search is truncated on containment measures (number of shared hashes) rather than similarity.

The test in here uses `--best-only` to demonstrate the behavior because it is easy, but the bug will also show up in other searches; this is the line of code in `sbtmh.py` that needs to be fixed:

```
if len(mins) and float(matches) / len(mins) >= threshold:
```

In this line, we also need to take into the number of mins in the search signature. This requires modification of the data stored at each internal node of the SBT.

Additionally, this PR:

* moves the search and gather code into a new module, `sourmash_lib.search`;
* adds `--traverse-directory` to `sourmash search` and `gather` (fixes #257);
* limits num results to 1 with `--best-only` (fixes #281);

This PR need some cleanup as well - the `gather` functionality should be revisited before a merge is requested - although it does pass all tests.

- [ ] Is it mergeable?
- [ ] `make test` Did it pass the tests?
- [ ] `make coverage` Is the new code covered?
- [ ] Did it change the command-line interface? Only additions are allowed
  without a major version increment. Changing file formats also requires a
  major version number increment.
- [ ] Was a spellchecker run on the source code and documentation after
  changes were made?

### CODE REVIEW

**Comment by ctb** [comment]
@luizirber your thoughts on the use of `max_n_below` on the SBT are requested! Briefly, I decorate the SBT internal nodes with a new `metadata` dictionary and store that dictionary in the SBT JSON file.

File: sourmash_lib/search.py [hunk 1]
```diff
@@ -0,0 +1,173 @@
...
+
+def gather_databases(query, databases, threshold_bp):
+    from sourmash_lib.sbtmh import SearchMinHashesFindBestIgnoreMaxHash
+
+    orig_query = query
+    orig_mins = orig_query.minhash.get_hashes()
+    query_ksize = query.minhash.ksize
+
+    # calculate the band size/resolution R for the genome
+    R_metagenome = sourmash_lib.MAX_HASH / float(orig_query.minhash.max_hash)
```
Comment (luizirber) [inline sourmash_lib/search.py]: We can use `get_scaled_for_max_hash` here

**Comment by luizirber** [comment]
Looking good so far!

**Comment by ctb** [comment]
This is substantial enough that maybe it can be merged now - it has tests for the critical bug that it fixes, and resolving merge conflicts is getting annoying. @luizirber thoughts?

**Comment by luizirber** [comment]
Seems like `tests/test-data/genome-s10+s11.fa.gz` is missing, @ctb

**Comment by ctb** [comment]
whoops! fixed. thx!

**Comment by ctb** [comment]
Huh. That's a weird py2.7 error....

**Comment by luizirber** [comment]
@ctb random scipy installation error is gone, but now the test is failing: https://travis-ci.org/dib-lab/sourmash/jobs/292597698#L630

**Comment by luizirber** [comment]
I can track down the error, pretty sure it is a division problem with py2...

**Comment by luizirber** [comment]
Yup: https://github.com/dib-lab/sourmash/pull/244/commits/432214087cde1e2ff20537d697dff894282a6681#diff-7669e75d60f9ec98d5284b0a0e7366c5R1

**Pull request merged**